### PR TITLE
Add string argument to jupyter.execSelectionInteractive for extensibility

### PR DESCRIPTION
**Pull request by andycraig**
For #1689

-   [x] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR).
-   [x] Title summarizes what is changing.
-   [x] Has a [news entry](https://github.com/Microsoft/vscode-jupyter/tree/main/news) file (remember to thank yourself!).
-   [x] Appropriate comments and documentation strings in the code.
-   [ ] Has sufficient logging.
-   [ ] Has telemetry for enhancements.
-   [x] Unit tests & system/integration tests are added/updated.
-   [ ] [Test plan](https://github.com/Microsoft/vscode-jupyter/blob/main/.github/test_plan.md) is updated as appropriate.
-   [x] ~[`package-lock.json`](https://github.com/Microsoft/vscode-jupyter/blob/main/package-lock.json) has been regenerated by running `npm install` (if dependencies have changed).~

Not sure about logging/telemetry/test plan items above.

This PR adds a string argument to `jupyter.execSelectionInteractive` for extensibility. This allows users to send custom text to the Interactive Window using keybindings, and (more importantly) allows other extensions to send custom text to the Interactive Window. Existing <kbd>Shift+Enter</kbd> behaviour of `jupyter.execSelectionInteractive` is unchanged.

**Demo**

I've added a unit test but this demo might make it clearer how this feature is intended to be used.

Add this to `keybindings.json`:

```json
    {
        "key": "alt+shift+a",
        "command":"jupyter.execSelectionInteractive",
        "args": "print('sent to Interactive Window by command')"
    }
```

Create a file `temp.py` with the following text:

```py
print('sent using Shift+Enter')
```

Enable setting `Jupyter: Send Selection to Interactive Window` (necessary for checking <kbd>Shift+Enter</kbd> behaviour)

Create an Interactive Window using command `Jupyter: Create Interactive Window`

Use <kbd>Alt+Shift+a</kbd>, observe that `print('sent to Interactive Window by command')` is sent to Interactive Window (new feature)

Use <kbd>Shift+Enter</kbd>, observe that `print('sent using Shift+Enter')` is sent to Interactive Window (preserving existing behaviour)

### CODE REVIEW

File: src/client/datascience/commands/commandRegistry.ts
Comment (rchiodo) [inline src/client/datascience/commands/commandRegistry.ts:84]: We usually use 'undefined' instead of null

**Review by rchiodo — CHANGES_REQUESTED** [review]
Looks good except for the 'null' vs 'undefined'.

**Comment by andycraig** [comment]
@rchiodo Thank you for the review! I've pushed a new commit changing 'null' to 'undefined'.

Apologies for the delay on the CLA, I'm still sorting out some paperwork for it.

**Review by rchiodo — APPROVED** [review]

**Review by DavidKutu — APPROVED** [review]

File: src/client/datascience/commands/commandRegistry.ts
Comment (andycraig) [reply src/client/datascience/commands/commandRegistry.ts:83]: Okay for me to make further changes after the PR's been approved? If so I'm happy to remove these redundant `= undefined`s.

Comment (rchiodo) [reply src/client/datascience/commands/commandRegistry.ts:83]: Yeah definitely. You also have to fix the linter errors before we can submit. 

The other failures - the VS code test failures - are a result of API changes from VS code affecting our tests and aren't related to your PR (so we can submit with them failing). But the linter errors have to be fixed.

Thanks

**Comment by rchiodo** [comment]
Oh and it looks like the CLA is still not recognized as being signed. We have to wait for that too.

File: src/client/datascience/commands/commandRegistry.ts
Comment (andycraig) [reply src/client/datascience/commands/commandRegistry.ts:83]: I tried removing the `= undefined`s but in both cases I get TypeScript errors:

undefined

**Comment by andycraig** [comment]
@rchiodo Looking at the CI linter message, the error seems to have occurred before the linters were run: 

```
Verify linting is turned on for changed files
Run ./.github/actions/check-ignore-list
Error: Files are being ignored that should be linted: src/test/datascience/editor-integration/codewatcher.unit.test.ts
```

In `.eslintrc.js`, `ignorePatterns` does include 
`'src/test/datascience/editor-integration/codewatcher.unit.test.ts'` (line 256). I changed that unit test file to add/update tests for a function I modified.

Should I modify `.eslintrc.js` to remove the unit test file from `ignorePatterns`?

**Comment by rchiodo** [comment]
> Should I modify .eslintrc.js to remove the unit test file from ignorePatterns?

Yep. That's what that check is essentially for. When we switched to Eslint, we didn't want to fix everything at once. So the CI system was updated to force ESlint only when a file was changed.

File: src/client/datascience/commands/commandRegistry.ts
Comment (rchiodo) [reply src/client/datascience/commands/commandRegistry.ts:83]: You have to fix the Commands.ts file so that the parameters for ExecSelectionInInteeractiveWindow match the value you're using.

Then the second error means you should change the (text: string | undefined) to (text?: string) because the argument is actually optional.

Comment (rchiodo) [reply src/client/datascience/commands/commandRegistry.ts:83]: See this spot here:
https://github.com/microsoft/vscode-jupyter/blob/ea611e5faf36c22abd2589085f03f81869365970/src/client/common/application/commands.ts#L35

**Comment by andycraig** [comment]
@rchiodo After removing some changed files from the lint ignore list I got lint errors 'Promises must be handled appropriately or explicitly marked as ignored with the `void` operator'. I fixed them by applying the suggested Code Actions to add `void` at the start of each of those lines. It made the lint errors go away but I don't have the background to judge whether those were definitely the correct fixes or not, so if you could check that commit to see that the changes look appropriate that would be great: 9dd9037

**Comment by rchiodo** [comment]
> @rchiodo After removing some changed files from the lint ignore list I got lint errors 'Promises must be handled appropriately or explicitly marked as ignored with the `void` operator'. I fixed them by applying the suggested Code Actions to add `void` at the start of each of those lines. It made the lint errors go away but I don't have the background to judge whether those were definitely the correct fixes or not, so if you could check that commit to see that the changes look appropriate that would be great: [9dd9037](https://github.com/microsoft/vscode-jupyter/commit/9dd90371ec791ffc25b28f0a14f46d771e347b91)

Yeah looks good to me.

**Comment by andycraig** [comment]
Great! Thank you very much for your reviews!

**Pull request merged**
